### PR TITLE
[LibOS] Return `-EINVAL` on encountering O_PATH flag in `open`/`openat`

### LINKED ISSUE #2538: Syscalls should fail with EBADF, but successfully goes through
## Description of the problem
Following syscalls should fail with EBADF, but they are successful in Graphene.
A file descriptor is obtained that can be used to perform operations  that act purely at the file descriptor level, the file itself is  not opened, the operations read(2), write(2), fchmod(2), fchown(2) fail with the error EBADF.

## Steps to reproduce
Execute attached files
open13_setup in linux environment and open13_run with graphene-direct
$ ./open13_setup
$ graphene-direct ./open13_run

## Expected results
open13      1  TPASS  :  read(2) failed with EBADF
open13      2  TPASS  :  write(2) failed with EBADF
open13      3  TPASS  :  fchmod(2) failed with EBADF
open13      4  TPASS  :  fchown(2) failed with EBADF
open13      5  TPASS  :  fgetxattr(2) failed with EBADF

## Actual results
open13     1 TFAIL : read(2) succeeded unexpectedly
open13     2 TFAIL : write(2) failed unexpectedly, expected EBADF: TEST_ERRNO=SUCCESS(0): Success
open13     3 TFAIL : fchmod(2) succeeded unexpectedly
open13     4 TFAIL : fchown(2) succeeded unexpectedly

## Labels
P0, bug
[open13.zip](https://github.com/oscarlab/graphene/files/6813957/open13.zip)

### PR DESCRIPTION
**Pull request by sahason**
## Description of the changes 
Fixes #2538. Return `-EINVAL` on encountering O_PATH flag in `open`/`openat`

## How to test this PR? 
Do the issue  repro steps mentioned in this PR #2538 .

---
This change is [Reviewable](https://reviewable.io/reviews/oscarlab/graphene/2614)

### CODE REVIEW

**Review by dimakuv — CHANGES_REQUESTED** [review]
> *[Reviewable](https://reviewable.io/reviews/oscarlab/graphene/2614)* status: 0 of 2 files reviewed, 1 unresolved discussion, not enough approvals from maintainers (3 more required), not enough approvals from different teams (2 more required, approved so far: ) (waiting on @sahason)

*[a discussion](https://reviewable.io/reviews/oscarlab/graphene/2614#-MgEtIB-5aG4KefcRCKn:-MgEtIB-5aG4KefcRCKo:b-70e68b) (no related file):*
This implementation is wrong. It "emulates" only what the `open13` LTP test expects, but doesn't adhere to the actual requirements of `O_PATH` flag (see `man 2 open`):

> Obtain  a  file  descriptor that can be used for two purposes: to indicate a location in the filesystem tree and to perform operations that act purely at the file descriptor level.  The file itself is not opened,  and other  file operations (e.g., read(2), write(2), fchmod(2), fchown(2), fgetxattr(2), ioctl(2), mmap(2)) fail with the error EBADF.

> Opening  a  file  or  directory  with the O_PATH flag requires no permissions on the object itself (but does require execute permission on the directories in the path prefix).

A simple test: create a file, remove its read permission (`chmod -r <file>`), and observe how Graphene fails to `open(O_PATH)` this file.

It's hard to say what the actual support for `O_PATH` means to Graphene. We should create the `shim_handle` object, but we shouldn't actually try to open the file (I don't see how this can easily be done in our current fs/chroot code). @pwmarcz Any thought on `O_PATH` support?

---

**Review by pwmarcz — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/oscarlab/graphene/2614)* status: 0 of 2 files reviewed, 1 unresolved discussion, not enough approvals from maintainers (3 more required), not enough approvals from different teams (2 more required, approved so far: ) (waiting on @dimakuv)

*[a discussion](https://reviewable.io/reviews/oscarlab/graphene/2614#-MgEtIB-5aG4KefcRCKn:-MgG9zIFDH0aBQJj1xch:b-5yapye) (no related file):*
<details><summary><i>Previously, dimakuv (Dmitrii Kuvaiskii) wrote…</i></summary><blockquote>

This implementation is wrong. It "emulates" only what the `open13` LTP test expects, but doesn't adhere to the actual requirements of `O_PATH` flag (see `man 2 open`):

> Obtain  a  file  descriptor that can be used for two purposes: to indicate a location in the filesystem tree and to perform operations that act purely at the file descriptor level.  The file itself is not opened,  and other  file operations (e.g., read(2), write(2), fchmod(2), fchown(2), fgetxattr(2), ioctl(2), mmap(2)) fail with the error EBADF.

> Opening  a  file  or  directory  with the O_PATH flag requires no permissions on the object itself (but does require execute permission on the directories in the path prefix).

A simple test: create a file, remove its read permission (`chmod -r <file>`), and observe how Graphene fails to `open(O_PATH)` this file.

It's hard to say what the actual support for `O_PATH` means to Graphene. We should create the `shim_handle` object, but we shouldn't actually try to open the file (I don't see how this can easily be done in our current fs/chroot code). @pwmarcz Any thought on `O_PATH` support?
</blockquote></details>

I don't think making the LTP test pass will be useful by itself. If we have a real application that uses it, we might look into how to make it work.

As for proper support: maybe this should be handled by another filesystem (same as we have "pipe", "socket" and "fifo", we could have a "path" filesystem, and probably a corresponding handle type (`hdl->type`)). Then it could implement callbacks that return `-EBADF` for most operations. I would first check how it's implemented in Linux, though.

However, if there is no use case for this, then I'd say it's not worth the complexity. Instead, we could explicitely *not* support the `O_PATH` flag and make `open/openat` return `-EINVAL` on encountering it.

---

**Comment by sahason** [comment]
Thanks @dimakuv @pwmarcz. Then I will update this PR by not supporting the O_PATH flag and make open/openat return -EINVAL on encountering it for now.

**Review by dimakuv — CHANGES_REQUESTED** [review]
> *[Reviewable](https://reviewable.io/reviews/oscarlab/graphene/2614)* status: 0 of 2 files reviewed, 2 unresolved discussions, not enough approvals from maintainers (3 more required), not enough approvals from different teams (2 more required, approved so far: ), "fixup! " found in commit messages' one-liners (waiting on @dimakuv and @sahason)

___
*[LibOS/shim/src/sys/shim_open.c, line 99 at r2](https://reviewable.io/reviews/oscarlab/graphene/2614#-MgL1aoHARj_lV8YWg1m:-MgL1aoIBCn-dmTnUTn5:br8l7xn) ([raw file](https://github.com/oscarlab/graphene/blob/6df1926eac97ff8cd6ab02f319020cd4b43e984d/LibOS/shim/src/sys/shim_open.c#L99)):*
> ```C
> 
> long shim_do_open(const char* file, int flags, mode_t mode) {
>     return shim_do_openat(AT_FDCWD, file, flags, mode);
> ```

Please also add this check here.

**Review by boryspoplawski — CHANGES_REQUESTED** [review]
> Reviewed 2 of 2 files at r2.
> *[Reviewable](https://reviewable.io/reviews/oscarlab/graphene/2614)* status: all files reviewed, 3 unresolved discussions, not enough approvals from maintainers (3 more required), not enough approvals from different teams (2 more required, approved so far: ), "fixup! " found in commit messages' one-liners (waiting on @dimakuv and @sahason)

*[a discussion](https://reviewable.io/reviews/oscarlab/graphene/2614#-MgNg8Nl4FmBruOSlXiK:-MgNg8Nl4FmBruOSlXiL:b9udt26) (no related file):*
Please update the commit message to match the current version of this PR and adhere to our contributing guidelines.

---
___
*[LibOS/shim/src/sys/shim_open.c, line 99 at r2](https://reviewable.io/reviews/oscarlab/graphene/2614#-MgL1aoHARj_lV8YWg1m:-MgNfJbl2u4fJNKNdl01:b-7np9z) ([raw file](https://github.com/oscarlab/graphene/blob/6df1926eac97ff8cd6ab02f319020cd4b43e984d/LibOS/shim/src/sys/shim_open.c#L99)):*
<details><summary><i>Previously, dimakuv (Dmitrii Kuvaiskii) wrote…</i></summary><blockquote>

Please also add this check here.
</blockquote></details>

No need, the check is in `shim_do_openat`...

**Review by sahason — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/oscarlab/graphene/2614)* status: all files reviewed, 3 unresolved discussions, not enough approvals from maintainers (3 more required), not enough approvals from different teams (2 more required, approved so far: ) (waiting on @boryspoplawski and @dimakuv)

*[a discussion](https://reviewable.io/reviews/oscarlab/graphene/2614#-MgNg8Nl4FmBruOSlXiK:-MgOu2qvBB8z-CwuHV7t:b-896fix) (no related file):*
<details><summary><i>Previously, boryspoplawski (Borys Popławski) wrote…</i></summary><blockquote>

Please update the commit message to match the current version of this PR and adhere to our contributing guidelines.
</blockquote></details>

Done.

---

**Review by dimakuv — CHANGES_REQUESTED** [review]
> Reviewed 1 of 2 files at r2.
> *[Reviewable](https://reviewable.io/reviews/oscarlab/graphene/2614)* status: all files reviewed, 1 unresolved discussion, not enough approvals from maintainers (3 more required), not enough approvals from different teams (2 more required, approved so far: ) (waiting on @boryspoplawski)

*[a discussion](https://reviewable.io/reviews/oscarlab/graphene/2614#-MgNg8Nl4FmBruOSlXiK:-MgPRake2Eg4vmUC6Mit:b-jl4ngc) (no related file):*
<details><summary><i>Previously, sahason wrote…</i></summary><blockquote>

Done.
</blockquote></details>

Could you also change the PR description (by clicking "Edit" on GitHub)? Explicitly mention that we return `-EINVAL` instead of trying to work with `O_PATH`.

Also, the commit message is too long, please shorten to this:
```
[LibOS] Return `-EINVAL` on encountering O_PATH flag in `open`/`openat`
```

---
___
*[LibOS/shim/src/sys/shim_open.c, line 99 at r2](https://reviewable.io/reviews/oscarlab/graphene/2614#-MgL1aoHARj_lV8YWg1m:-MgPRT2j6Pk5QZszL05e:b-f32not) ([raw file](https://github.com/oscarlab/graphene/blob/6df1926eac97ff8cd6ab02f319020cd4b43e984d/LibOS/shim/src/sys/shim_open.c#L99)):*
<details><summary><i>Previously, boryspoplawski (Borys Popławski) wrote…</i></summary><blockquote>

No need, the check is in `shim_do_openat`...
</blockquote></details>

True. Me being stupid.

**Review by sahason — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/oscarlab/graphene/2614)* status: all files reviewed, 1 unresolved discussion, not enough approvals from maintainers (3 more required), not enough approvals from different teams (2 more required, approved so far: ) (waiting on @boryspoplawski and @dimakuv)

*[a discussion](https://reviewable.io/reviews/oscarlab/graphene/2614#-MgNg8Nl4FmBruOSlXiK:-MgPYtUBApKGG-6yMDcv:b-896fix) (no related file):*
<details><summary><i>Previously, dimakuv (Dmitrii Kuvaiskii) wrote…</i></summary><blockquote>

Could you also change the PR description (by clicking "Edit" on GitHub)? Explicitly mention that we return `-EINVAL` instead of trying to work with `O_PATH`.

Also, the commit message is too long, please shorten to this:
```
[LibOS] Return `-EINVAL` on encountering O_PATH flag in `open`/`openat`
```
</blockquote></details>

Done.

---

**Review by dimakuv — APPROVED** [review]
> *[Reviewable](https://reviewable.io/reviews/oscarlab/graphene/2614)* status: all files reviewed, 1 unresolved discussion, not enough approvals from maintainers (2 more required), not enough approvals from different teams (1 more required, approved so far: Intel) (waiting on @boryspoplawski)

**Review by mkow — CHANGES_REQUESTED** [review]
> Reviewed 1 of 2 files at r2.
> *[Reviewable](https://reviewable.io/reviews/oscarlab/graphene/2614)* status: all files reviewed, 2 unresolved discussions, not enough approvals from maintainers (2 more required), not enough approvals from different teams (1 more required, approved so far: Intel) (waiting on @boryspoplawski and @sahason)

*[a discussion](https://reviewable.io/reviews/oscarlab/graphene/2614#-MgQx64P4ZODJzC9DYbD:-MgQx64P4ZODJzC9DYbE:b-ca6k74) (no related file):*
Hmm, why does this PR include a change from another PR in the first commit, and then reverts it in the second?

---

**Review by mkow — CHANGES_REQUESTED** [review]
> *[Reviewable](https://reviewable.io/reviews/oscarlab/graphene/2614)* status: all files reviewed, 3 unresolved discussions, not enough approvals from maintainers (2 more required), not enough approvals from different teams (1 more required, approved so far: Intel) (waiting on @boryspoplawski and @sahason)

*[a discussion](https://reviewable.io/reviews/oscarlab/graphene/2614#-MgQxOgo76Edc2QHiv8w:-MgQxOgp1JvwWBiN3P3I:b7yjuet) (no related file):*
Also, the second commit is missing the sign-off, so the PR got blocked by the DCO check.

---

**Review by sahason — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/oscarlab/graphene/2614)* status: all files reviewed, 3 unresolved discussions, not enough approvals from maintainers (2 more required), not enough approvals from different teams (2 more required, approved so far: ) (waiting on @boryspoplawski and @mkow)

*[a discussion](https://reviewable.io/reviews/oscarlab/graphene/2614#-MgQx64P4ZODJzC9DYbD:-MgdUTG10id2_3xaM0BX:b-kml569) (no related file):*
<details><summary><i>Previously, mkow (Michał Kowalczyk) wrote…</i></summary><blockquote>

Hmm, why does this PR include a change from another PR in the first commit, and then reverts it in the second?
</blockquote></details>

At first I have submitted this PR to handle what the open13 LTP test expects which seems to be not appropriate. Further as per the discussion I have made an update to return -EINVAL on encountering O_PATH flag in open/openat.

---
*[a discussion](https://reviewable.io/reviews/oscarlab/graphene/2614#-MgQxOgo76Edc2QHiv8w:-MgdUP3g2a4kx-zBYwEt:b-un0g5e) (no related file):*
<details><summary><i>Previously, mkow (Michał Kowalczyk) wrote…</i></summary><blockquote>

Also, the second commit is missing the sign-off, so the PR got blocked by the DCO check.
</blockquote></details>

My mistake. I have updated it.

---

**Review by dimakuv — APPROVED** [review]
> Reviewed 1 of 1 files at r6.
> *[Reviewable](https://reviewable.io/reviews/oscarlab/graphene/2614)* status: all files reviewed, 3 unresolved discussions, not enough approvals from maintainers (1 more required), not enough approvals from different teams (1 more required, approved so far: Intel) (waiting on @boryspoplawski and @mkow)

*[a discussion](https://reviewable.io/reviews/oscarlab/graphene/2614#-MgQx64P4ZODJzC9DYbD:-MgdYbBa3LRWG_NypxiK:b-i7t0l9) (no related file):*
<details><summary><i>Previously, sahason wrote…</i></summary><blockquote>

At first I have submitted this PR to handle what the open13 LTP test expects which seems to be not appropriate. Further as per the discussion I have made an update to return -EINVAL on encountering O_PATH flag in open/openat.
</blockquote></details>

DONE. I updated this PR to have only one (second) commit. The issue here was that the initial version of this PR fixed the problem in a wrong way, so @sahason fixed the problem in another way in the second commit (effectively reverting the changes in the first commit). Our guidelines are fuzzy on what to do in such a situation: maybe he should have created another PR completely and close this one. Or maybe he should have `git rebase` these two commits to squash them into one, but our guidelines explicitly forbit to rebase (only in super-rare cases which is... nobody knows what these are :)).

Anyway, to avoid further confusion, I squashed the two commits into one and rebased them (using `git git rebase -i HEAD~2 --autosquash` and choosing to `edit` the first commit and `squash` the second commit, which allows to merge the two). @sahason Hope you don't mind.

---

**Comment by dimakuv** [comment]
Jenkins, test this please

**Review by boryspoplawski — DISMISSED** [review]
> Reviewed 1 of 1 files at r6.
> *[Reviewable](https://reviewable.io/reviews/oscarlab/graphene/2614)* status: all files reviewed, 2 unresolved discussions (waiting on @mkow)

**Review by mkow — DISMISSED** [review]
> Reviewed 1 of 2 files at r2, 1 of 1 files at r6.
> *[Reviewable](https://reviewable.io/reviews/oscarlab/graphene/2614)* status: :shipit: complete! all files reviewed, all discussions resolved

**Comment by mkow** [comment]
Jenkins, retest this please

**Review by boryspoplawski — APPROVED** [review]
> *[Reviewable](https://reviewable.io/reviews/oscarlab/graphene/2614)* status: all files reviewed, all discussions resolved, not enough approvals from maintainers (1 more required), not enough approvals from different teams (1 more required, approved so far: ITL)

**Review by dimakuv — APPROVED** [review]
> *[Reviewable](https://reviewable.io/reviews/oscarlab/graphene/2614)* status: :shipit: complete! all files reviewed, all discussions resolved

**Pull request merged**
